### PR TITLE
refactor: centralize Android NFC method exception handling

### LINKED ISSUE #168: bug: MIFARE: unhandled security exceptions
The Android NFC interface throws SecurityExceptions for the MIFARE Classic read/write/authenticate methods when an operation is called but the tag has left the field.

This are currently unhandled by the library and results in app crashes.

### PR DESCRIPTION
**Pull request by knthm**
Reduces code duplication and improves readability.

Fixes #168

### CODE REVIEW

**Comment by Harry-Chen** [comment]
Thanks! I am actually thinking about some more elegant ways to handle all possible exceptions. Something like:

```kotlin
fun handleException(ex: Exception, result: Result?, desc: String) {
    Log.e(TAG, "$desc error", ex)
    val excMessage = ex.localizedMessage
    when (ex) {
        is IOException -> result?.error("500", "Communication error", excMessage )
        is SecurityException -> result?.error("503", "Tag already removed", excMessage)
        // some more types
        else -> result?.error("510", "Unhandled error", excMessage) // I choose this code just randomly.
    }
}
```

Then it could be called from all `catch` blocks. What's your opinion on this?

**Comment by knthm** [comment]
With the current amount of `catch` blocks per operation I was thinking along the same lines, but thought it safer at first to maintain the same pattern.

I suppose refactoring the exception handling wouldn't take too much time and would definitely make things a lot cleaner.

My main concern is I can't physically test the other tag types to ensure the behavior triggering the exception doesn't differ slightly.

**Comment by Harry-Chen** [comment]
> My main concern is I can't physically test the other tag types to ensure the behavior triggering the exception doesn't differ slightly.

AFAIK Android NFC APIs mainly throw:

* `IOException` and `SecurityException` in every IO-related API
* `IllegalArgumentException` in `transceive`
* `FormatException` in NDEF-related APIs

There are additional `InvocationTargetException` and `NoSuchMethodException` because reflection is used to simplify `transceive`.

Seems it's safe to refactor, since no two different APIs throws the same exceptions except `IOException` and `SecurityException`.

**Comment by knthm** [comment]
Good points, I went ahead and offloaded it into more or less the copy-pasted version of your suggestion :)

**Comment by Harry-Chen** [comment]
Yeah! This looks much better now. I just found that it could be merged into `runOnNfcThread`, like:

```kotlin
private fun runOnNfcThread(result: Result, desc: String, fn: () -> Unit) {
    val wrappedFn = Runnable {
        try {
            fn()
        } catch (ex: Exception) {
            handleException(ex, result, desc)
        }
    }
    if (!nfcHandler.post(fn)) {
        result.error("500", "Failed to post job to NFC Handler thread.", null)
    }
}
```

Then:

- `result` could be non-null.
- the `catch` here could be absorbed: https://github.com/nfcim/flutter_nfc_kit/blob/27fe9ba522d78702b02eba0729bf7e9ba9998608/android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt#L595-L600 

By the way, there is an extra space at: https://github.com/nfcim/flutter_nfc_kit/blob/27fe9ba522d78702b02eba0729bf7e9ba9998608/android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt#L70

Will you please make the changes, and I'm really happy to merge.

**Comment by knthm** [comment]
> I just found that it could be merged into `runOnNfcThread`

Oh! That's a lot better, agreed.

I ended up moving `handleException` into `runOnNfcThread`, since they share the exact same use case: interaction with the OS's NFC methods that may block and may throw and exception.

File: android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt [hunk 1]
```diff
@@ -614,11 +560,8 @@ class FlutterNfcKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             result.success(jsonResult.toString())
         }
 
-        try {
+        runOnNfcThread(result, "Enable reader mode") {
```
Comment (Harry-Chen) [inline android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt:563]: You don't need. `pollTag` is called by `handleMethodCall` in `runOnNfcThread`, so just let the error propagate outside the function.

Comment (knthm) [reply android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt:563]: You're right, removed.

**Pull request merged**